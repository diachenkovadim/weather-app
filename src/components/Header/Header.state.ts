import { useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

import { apiServices } from 'services/apiService';
import { fetchWeatherData, removeError } from 'services/redux/weatherReducer';
import { useAppDispatch, useAppSelector } from 'hooks/useRedux';
import useDebounce from 'hooks/useDebounce';

import { ROUTES } from 'constants/routes';
import { useSnackbar } from 'notistack';

interface IGeoAPIresponse {
  name: string;
  local_names: {
    [key: string]: string;
    feature_name: string;
  }[];
  lat: number;
  lon: number;
  country: string;
  state?: string;
}

export const useHeaderState = () => {
  const [isAutocompleteOpen, setIsAutocompleteOpen] = useState(false);
  const [searchValue, setSearchValue] = useState('');
  const [cities, setCities] = useState<IGeoAPIresponse[]>([]);
  const [selectedCity, setSelectedCity] = useState<IGeoAPIresponse | null>(
    null
  );
  const { enqueueSnackbar } = useSnackbar();

  const { isError, errorMessage } = useAppSelector((state) => state.weather);

  const onRemoveErroHandler = () => dispatch(removeError());

  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const { pathname } = useLocation();

  const onOpenAutocompleteHandler = () => setIsAutocompleteOpen(true);
  const onCloseAutocompleteHandler = () => setIsAutocompleteOpen(false);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchValue(event.target.value);
    setSelectedCity(null);
  };

  const onAutocompleteChangeHandler = (
    event: React.SyntheticEvent<Element, Event>,
    value: IGeoAPIresponse | null
  ) => {
    setSelectedCity(value);
    onClickItemHandler(value?.lat || 0, value?.lon || 0);
  };

  const onClickItemHandler = (lat: number, lon: number) => {
    dispatch(fetchWeatherData({ lat, lon }));
    onCloseAutocompleteHandler();
    setCities([]);
    setSearchValue('');
    pathname !== ROUTES.home && navigate(ROUTES.home);
  };

  const debouncedSearchValue = useDebounce(searchValue);

  const onGetCityHandler = async () => {
    try {
      if (searchValue) {
        const { data }: { data: IGeoAPIresponse[] } =
          await apiServices.fetchCityGeoData({
            q: debouncedSearchValue,
            limit: 5,
          });
        setCities(data);
      } else {
        setCities([]);
      }
    } catch (error) {
      enqueueSnackbar({ variant: 'error', message: 'Failed to fetch cities' });
      console.log(error);
    }
  };

  return {
    onOpenAutocompleteHandler,
    onCloseAutocompleteHandler,
    onGetCityHandler,
    handleChange,
    onClickItemHandler,
    onRemoveErroHandler,
    enqueueSnackbar,
    onAutocompleteChangeHandler,
    selectedCity,
    errorMessage,
    isError,
    pathname,
    cities,
    debouncedSearchValue,
    searchValue,
    isAutocompleteOpen,
  };
};
