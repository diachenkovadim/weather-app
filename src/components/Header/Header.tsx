import { FC, useEffect } from 'react';
import { AppBar, Autocomplete, Stack, TextField, Toolbar } from '@mui/material';

import { useHeaderState } from './Header.state';

import { HEADER_CONFIG } from 'constants/config';

export const Header: FC = () => {
  const {
    onCloseAutocompleteHandler,
    onOpenAutocompleteHandler,
    handleChange,
    onGetCityHandler,
    onClickItemHandler,
    isAutocompleteOpen,
    cities,
    debouncedSearchValue,
    pathname,
    isError,
    errorMessage,
    onRemoveErroHandler,
    enqueueSnackbar,
    selectedCity,
    onAutocompleteChangeHandler,
  } = useHeaderState();

  useEffect(() => {
    onGetCityHandler();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [debouncedSearchValue]);

  useEffect(() => {
    isAutocompleteOpen && onCloseAutocompleteHandler();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [pathname]);

  useEffect(() => {
    isError &&
      enqueueSnackbar({
        variant: 'error',
        message: errorMessage,
      });
    return () => {
      onRemoveErroHandler();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isError]);

  return (
    <AppBar position="static" color="inherit">
      <Toolbar
        sx={{
          justifyContent: 'space-between',
          height: {
            xs: HEADER_CONFIG.H_MOBILE,
            md: HEADER_CONFIG.H_MAIN,
          },
        }}
      >
        <Autocomplete
          data-testid="autocomplete"
          fullWidth
          size="small"
          value={selectedCity}
          open={isAutocompleteOpen}
          onOpen={onOpenAutocompleteHandler}
          clearOnEscape
          onBlur={onCloseAutocompleteHandler}
          onChange={onAutocompleteChangeHandler}
          getOptionLabel={(data) =>
            `${data.name}, ${data.country}. Latitude ${data.lat} - Longitude ${data.lon}`
          }
          renderInput={(params) => (
            <TextField
              {...params}
              data-testid="input"
              label="Enter city name"
              fullWidth
              onChange={handleChange}
            />
          )}
          options={cities}
          renderOption={(props, option, state) => {
            const onClick = () => {
              onClickItemHandler(option.lat, option.lon);
              state.inputValue = '';
            };
            return (
              <li
                {...props}
                key={`${option.lat}${option.lon}`}
                onClick={onClick}
              >
                <Stack direction="row" spacing={1}>
                  <div>{`${option.name}, ${option.country}. Latitude ${option.lat} - Longitude ${option.lon}`}</div>
                </Stack>
              </li>
            );
          }}
          filterOptions={(x) => x}
        />
      </Toolbar>
    </AppBar>
  );
};
