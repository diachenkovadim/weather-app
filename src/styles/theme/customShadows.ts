import { alpha } from '@mui/material/styles';

import { palette } from './palette';

interface CustomShadowOptions {
  primary: string;
  info: string;
  card: string;
}

declare module '@mui/material/styles' {
  interface Theme {
    customShadows: CustomShadowOptions;
  }
  interface ThemeOptions {
    customShadows?: CustomShadowOptions;
  }
}

const themeColor = palette();

const LIGHT_MODE = themeColor.grey[500];

const createShadow = (color: string) => {
  return {
    primary: `0 8px 16px 0 ${alpha(palette().primary.main, 0.24)}`,
    info: `0 8px 16px 0 ${alpha(palette().info.main, 0.24)}`,
    card: `0 0 2px 0 ${alpha(color, 0.2)}, 0 12px 24px -4px ${alpha(
      color,
      0.12
    )}`,
  };
};

export const customShadows = () => createShadow(LIGHT_MODE);
