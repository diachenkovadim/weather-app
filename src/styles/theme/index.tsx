import { useMemo } from 'react';

import { CssBaseline } from '@mui/material';
import {
  ThemeProvider as MUIThemeProvider,
  StyledEngineProvider,
  ThemeOptions,
  createTheme,
} from '@mui/material/styles';

import { customShadows } from './customShadows';
import { GlobalStyles } from './globalStyles';
import { palette } from './palette';
import { shadows } from './shadows';
import { typography } from './typography';

type TProps = {
  children: React.ReactNode;
};

export const ThemeProvider = ({ children }: TProps) => {
  const themeOptions: ThemeOptions = useMemo(
    () => ({
      palette: palette(),
      typography,
      shape: { borderRadius: 8 },
      shadows: shadows(),
      customShadows: customShadows(),
    }),
    []
  );

  const theme = createTheme(themeOptions);

  return (
    <StyledEngineProvider injectFirst>
      <MUIThemeProvider theme={theme}>
        <CssBaseline />
        <GlobalStyles />
        {children}
      </MUIThemeProvider>
    </StyledEngineProvider>
  );
};
