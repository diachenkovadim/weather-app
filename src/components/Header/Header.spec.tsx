import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Header } from './Header';
import { useHeaderState } from './Header.state';

import { ThemeProvider } from 'styles/theme';

jest.mock('./Header.state');

jest.mock('services/apiService/apiService', () => ({
  apiServices: {
    fetchCityGeoData: jest.fn(),
  },
}));

const mockOnGetCityHandlerMock = jest.fn();
const mockonClickItemHandler = jest.fn();
const baseConfig = {
  onOpenAutocompleteHandler: jest.fn(),
  onCloseAutocompleteHandler: jest.fn(),
  onGetCityHandler: mockOnGetCityHandlerMock,
  handleChange: jest.fn(),
  onClickItemHandler: mockonClickItemHandler,
  onRemoveErroHandler: jest.fn(),
  enqueueSnackbar: jest.fn(),
  onAutocompleteChangeHandler: jest.fn(),
  selectedCity: null,
  errorMessage: '',
  isError: false,
  pathname: '/',
  cities: [],
  debouncedSearchValue: '',
  searchValue: '',
  isAutocompleteOpen: true,
};

describe('Header', () => {
  const mockUseHeaderState = useHeaderState as jest.MockedFunction<
    typeof useHeaderState
  >;

  beforeEach(() => {
    mockUseHeaderState.mockReturnValue(baseConfig);
  });
  afterEach(() => {
    mockUseHeaderState.mockReset();
  });

  it('should render the header', () => {
    const view = render(
      <ThemeProvider>
        <Header />
      </ThemeProvider>
    );
    expect(view).toMatchSnapshot();
    expect(screen.getByLabelText('Enter city name')).toBeInTheDocument();
  });

  it('calls onGetCityHandler when search value changes', () => {
    render(
      <ThemeProvider>
        <Header />
      </ThemeProvider>
    );

    const searchInput = screen.getByLabelText(/Enter city name/i);
    fireEvent.change(searchInput, { target: { value: 'New York' } });
    expect(mockOnGetCityHandlerMock).toHaveBeenCalled();
  });
  it('renders the autocomplete options', async () => {
    const cities = [
      {
        name: 'New York',
        lat: 40,
        lon: 50,
        country: 'US',
        local_names: [],
      },
      {
        name: 'Los Angeles',
        lat: 34.0522,
        lon: -118.2437,
        country: 'US',
        local_names: [],
      },
    ];

    mockUseHeaderState.mockReturnValue({
      ...baseConfig,
      cities,
    });

    render(
      <ThemeProvider>
        <Header />
      </ThemeProvider>
    );

    const newYorkOption = await screen.findByText(/new york/i);
    expect(newYorkOption).toBeInTheDocument();
    userEvent.click(newYorkOption);
    expect(mockonClickItemHandler).toHaveBeenCalled();

    expect(await screen.findByText(/new york/i)).toBeInTheDocument();
    expect(await screen.findByText(/Los Angeles/i)).toBeInTheDocument();
  });
});
