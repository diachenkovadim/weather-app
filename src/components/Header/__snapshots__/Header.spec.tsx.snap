// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render the header 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorInherit MuiAppBar-positionStatic css-k30kuk-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-cwgnc7-MuiToolbar-root"
        >
          <div
            class="MuiAutocomplete-root Mui-expanded MuiAutocomplete-fullWidth MuiAutocomplete-hasPopupIcon css-f34qqq-MuiAutocomplete-root"
            data-testid="autocomplete"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              data-testid="input"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-6v1fmu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                Enter city name
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-182rbj6-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-invalid="false"
                  autocapitalize="none"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-mf0znp-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  role="combobox"
                  spellcheck="false"
                  type="text"
                  value=""
                />
                <div
                  class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                >
                  <button
                    aria-label="Close"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator MuiAutocomplete-popupIndicatorOpen css-142fjro-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                    tabindex="-1"
                    title="Close"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="ArrowDropDownIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Enter city name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </header>
    </div>
    <div
      class="MuiAutocomplete-popper css-1s5lphu-MuiPopper-root-MuiAutocomplete-popper MuiPopper-root"
      data-popper-escaped=""
      data-popper-placement="bottom"
      data-popper-reference-hidden=""
      role="presentation"
      style="position: absolute; top: 0px; left: 0px; width: 0px; margin: 0px; transform: translate(0px, 0px);"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAutocomplete-paper css-nsqo2c-MuiPaper-root-MuiAutocomplete-paper"
      >
        <div
          class="MuiAutocomplete-noOptions css-1fyfb1s-MuiAutocomplete-noOptions"
          role="presentation"
        >
          No options
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorInherit MuiAppBar-positionStatic css-k30kuk-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-cwgnc7-MuiToolbar-root"
      >
        <div
          class="MuiAutocomplete-root Mui-expanded MuiAutocomplete-fullWidth MuiAutocomplete-hasPopupIcon css-f34qqq-MuiAutocomplete-root"
          data-testid="autocomplete"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-testid="input"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-6v1fmu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r0:"
              id=":r0:-label"
            >
              Enter city name
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-182rbj6-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-mf0znp-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                role="combobox"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Close"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator MuiAutocomplete-popupIndicatorOpen css-142fjro-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Close"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Enter city name
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
